include/revng/Pipeline/Step.h

```diff
@@ -0,0 +1,387 @@
+void invokeImpl(const Context &Ctx,
+                Pipe &Enf,
+                void (Pipe::*F)(const Context &, Args...),
+                Containers &Containers,
+                const StringArrayRef &ArgsNames,
+                const std::integer_sequence<size_t, S...> &) {
+  (Enf
+   .*F)(Ctx,
+        getContainerFromName<std::decay_t<Args>>(Containers, ArgsNames[S])...);
+}
+
+template<typename Pipe, typename... Args, size_t... S>
+void invokeImpl(Pipe &Enf,
+                void (Pipe::*F)(Args...),
+                Containers &Containers,
+                const StringArrayRef &ArgsNames,
+                const std::integer_sequence<size_t, S...> &) {
+  (Enf
+   .*F)(getContainerFromName<std::decay_t<Args>>(Containers, ArgsNames[S])...);
+}
+
+///
+/// Invokes the F member function on the Pipe Enf passing as nth argument
+/// the container with the name equal to the nth element of ArgsNames
+///
+template<typename Pipe, typename... Args>
+void invokeEnforcingFunction(Context &,
+                             Pipe &Enf,
+                             void (Pipe::*F)(Args...),
+                             Containers &Containers,
+                             const llvm::ArrayRef<std::string> &ArgsNames) {
+  constexpr auto
+    Indexes = std::make_integer_sequence<size_t, sizeof...(Args)>();
+  revng_assert(sizeof...(Args) == ArgsNames.size());
+  invokeImpl(Enf, F, Containers, ArgsNames, Indexes);
+}
+
+template<typename Pipe, typename... Args>
+void invokeEnforcingFunction(Context &Ctx,
+                             Pipe &Enf,
+                             void (Pipe::*F)(Context &, Args...),
+                             Containers &Containers,
+                             const llvm::ArrayRef<std::string> &ArgsNames) {
+  constexpr auto
+    Indexes = std::make_integer_sequence<size_t, sizeof...(Args)>();
+  revng_assert(sizeof...(Args) == ArgsNames.size());
+  invokeImpl(Ctx, Enf, F, Containers, ArgsNames, Indexes);
+}
+
+template<typename Pipe, typename... Args>
+void invokeEnforcingFunction(Context &Ctx,
+                             Pipe &Enf,
+                             void (Pipe::*F)(const Context &, Args...),
+                             Containers &Containers,
+                             const llvm::ArrayRef<std::string> &ArgsNames) {
+  constexpr auto
+    Indexes = std::make_integer_sequence<size_t, sizeof...(Args)>();
+  revng_assert(sizeof...(Args) == ArgsNames.size());
+  invokeImpl(Ctx, Enf, F, Containers, ArgsNames, Indexes);
+}
+
+class PipeWrapperBaseImpl {
+public:
+  virtual void run(Context &Ctx, Containers &Containers) = 0;
+  virtual ContainersStatus
+  getRequirements(const ContainersStatus &Target) const = 0;
+  virtual ContainersStatus deduceResults(ContainersStatus &Target) const = 0;
+  virtual ~PipeWrapperBaseImpl() = default;
+  virtual std::unique_ptr<PipeWrapperBaseImpl> clone() const = 0;
+  virtual std::vector<std::string> getRunningContainersNames() const = 0;
```

Comment:
What exactly is a *running container*? Would it be appropriate to rename the `Containers` class to `RunningContainers[Set]`?

Reply:
Containers is both the "Backing Containers" and the "Running Containers" since the class is the same for both usages.

Reply:
Can you define *backing container* vs *running container*?
Backing is associated to the end of a step, while running is the one used while executing a step?

Reply:
yes